internal-api/src/main/java/datadog/trace/bootstrap/config/provider/StableConfigSource.java

```
@@ -0,0 +1,63 @@
+package datadog.trace.bootstrap.config.provider;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.InputStream;
+import java.util.HashMap;
+
+import static datadog.trace.util.Strings.propertyNameToEnvironmentVariableName;
+
+import datadog.trace.api.ConfigOrigin;
+import datadog.trace.bootstrap.config.provider.ConfigProvider;
+
+import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.LoaderOptions;
+import org.yaml.snakeyaml.constructor.SafeConstructor;
+
+final public class StableConfigSource extends ConfigProvider.Source {
+  static final String MANAGED_STABLE_CONFIGURATION_PATH = "/etc/datadog-agent/managed/datadog-agent/stable/datadog_apm.yaml";
+  static final String LOCAL_STABLE_CONFIGURATION_PATH = "/etc/datadog-agent/datadog_apm.yaml";
+
+  final ConfigOrigin fileOrigin;
+  HashMap<String, String> configuration;
+
+  StableConfigSource(String file, ConfigOrigin origin) {
+    this.fileOrigin = origin;
+    try {
+      configuration = parseStableConfig(file);
+    } catch (Exception e) {
+      configuration = new HashMap();
```

You should still be able to declare `configuration` as `final`, javac is smart enough to figure out only a single assignment happens here. For an empty map (if you change the declared type to `Map` instead of `HashMap`) you can also use `Map.of()` or `Collections.emptyMap()`. Finally, we should at least report the exception if the file exists but is not readable/parseable.

---

+1